### PR TITLE
[CALCITE] Script parsing case statement

### PR DESCRIPTION
**Pull request by ArthurInTheShell**
Implement the feature for parsing case statement in store procedure.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.calcite.util.ImmutableNullableList;
+
+import java.util.List;
+import java.util.Objects;
+/**
+ * Parse tree for a {@code SqlCaseStmt}.
+ */
+public class SqlCaseStmt extends SqlCall {
+  public static final SqlSpecialOperator OPERATOR =
+      new SqlSpecialOperator("CASE", SqlKind.CONDITIONAL_STATEMENT);
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java:29]: Should we use a more specific SqlKind for CONDITIONAL_STATEMENT (same for the existing SqlIfStmt type)?

File: core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    return OPERATOR;
+  }
+
+  @Override public List<SqlNode> getOperandList() {
+    return ImmutableNullableList.of(conditionalStmtListPairs, elseStmtList);
+  }
+
+  @Override public void unparse(final SqlWriter writer, final int leftPrec,
+      final int rightPrec) {
+    for (SqlNode pair: conditionalStmtListPairs) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java:60]: nit: Space before ":".

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl [hunk 3]
```diff
@@ -4825,7 +4825,11 @@ SqlNode ConditionalStmt() :
     final SqlNode e;
 }
 {
-    e = IfStmt()
+    (
+        e = IfStmt()
```
Comment (dasch-1) [inline parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4829]: nit: Order alphabetically.

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl [hunk 4]
```diff
@@ -4853,6 +4857,36 @@ SqlIfStmt IfStmt() :
     }
 }
 
+SqlCaseStmt CaseStmt() :
+{
+    SqlNode firstOperand = null;
+    SqlNode e = null;
+    final SqlNodeList conditionMultiStmtList = new SqlNodeList(getPos());
+    final SqlStatementList elseMultiStmtList = new SqlStatementList(getPos());
```
Comment (dasch-1) [inline parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4865]: nit: Line wrap

Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4865]: This line ends at 79 character, which is not against the style we are using (wrap at 80 character).

File: core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java (hunk 1 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java]: Added in https://github.com/googleinterns/calcite/pull/263/commits/59ec4c1cd314b7c1b1161ac20c8f9361d1c3c2ec

File: core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java (hunk 2 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlCaseStmt.java]: Fixed in https://github.com/googleinterns/calcite/pull/263/commits/59ec4c1cd314b7c1b1161ac20c8f9361d1c3c2ec

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl (hunk 3 above)
Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl]: Fixed in https://github.com/googleinterns/calcite/pull/263/commits/59ec4c1cd314b7c1b1161ac20c8f9361d1c3c2ec

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl (hunk 4 above)
Comment (dasch-1) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4865]: You're right - looks good.

File: parsing/dialects/intermediate/dialects/dialect1/src/test/java/org/apache/calcite/test/Dialect1ParserTest.java [hunk 5]
```diff
@@ -4593,15 +4605,84 @@ private void checkExNotNull(SqlParserUtil.StringAndPos sap, Throwable thrown) {
...
+    final String expected =
+        "CREATE PROCEDURE `FOO` (IN `BEE` INTEGER)\n"
+            + "CASE `BEE` WHEN 2 THEN SELECT `BAR`;\n"
+            + "WHEN 3 THEN SELECT `ABC`;\n"
+            + "ELSE SELECT `CDE`;\n"
+            + "END CASE";
+    sql(sql).ok(expected);
+  }
+
+  @Test public void testCaseStmtWithConditionalExpressionBase() {
```
Comment (lisayin) [inline parsing/dialects/intermediate/dialects/dialect1/src/test/java/org/apache/calcite/test/Dialect1ParserTest.java:4643]: Can you nest case statements?

Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/src/test/java/org/apache/calcite/test/Dialect1ParserTest.java:4643]: Yes, case statements can be nested. I've added a test case at https://github.com/googleinterns/calcite/pull/263/commits/3d8e35213af09f1537fd4d58f187acc458f4c37f parsing a nested case statement.

**Pull request merged**